Mathlib/Data/Matroid/Rank/ENat.lean

```diff
@@ -351,6 +352,188 @@ lemma IsRkFinite.eRk_lt_top (h : M.IsRkFinite X) : M.eRk X < ⊤ :=
+    (hr : M.eRk Y ≤ M.eRk X) : M.closure X = M.closure Y := by
+  obtain ⟨I, hI⟩ := M.exists_isBasis' X
+  obtain ⟨J, hJ, hIJ⟩ := hI.indep.subset_isBasis'_of_subset (hI.subset.trans hXY)
+  rw [hI.eRk_eq_encard, hJ.eRk_eq_encard] at hr
+  rw [← closure_inter_ground, ← M.closure_inter_ground Y,
+    ← hI.isBasis_inter_ground.closure_eq_closure,
+    ← hJ.isBasis_inter_ground.closure_eq_closure, Finite.eq_of_subset_of_encard_le
+      (hI.indep.finite_of_subset_isRkFinite hI.subset hX) hIJ hr]
+
+/-! ### Insertion -/
+
+lemma eRk_insert_le_add_one (M : Matroid α) (e : α) (X : Set α) :
+    M.eRk (insert e X) ≤ M.eRk X + 1 :=
+  union_singleton ▸ (M.eRk_union_le_eRk_add_eRk _ _).trans <|
+    add_le_add_left (by simpa using M.eRk_le_encard {e}) _
+
+lemma eRk_insert_eq_add_one (he : e ∈ M.E \ M.closure X) : M.eRk (insert e X) = M.eRk X + 1 := by
+  obtain ⟨I, hI⟩ := M.exists_isBasis' X
+  rw [← hI.closure_eq_closure, mem_diff, hI.indep.mem_closure_iff', not_and] at he
+  rw [← eRk_closure_eq, ← closure_insert_congr_right hI.closure_eq_closure, hI.eRk_eq_encard,
+    eRk_closure_eq, Indep.eRk_eq_encard (by tauto), encard_insert_of_not_mem (by tauto)]
+
+lemma exists_eRk_insert_eq_add_one_of_lt (h : M.eRk X < M.eRk Z) :
+    ∃ z ∈ Z \ X, M.eRk (insert z X) = M.eRk X + 1 := by
+  by_cases hz : Z ∩ M.E ⊆ M.closure X
+  · exact False.elim <| h.not_le <| by simpa using M.eRk_mono hz
+  obtain ⟨e, ⟨⟨heZ, heE⟩, heX⟩⟩ := not_subset.1 hz
+  exact ⟨e, ⟨heZ, fun heX' ↦ heX (mem_closure_of_mem' _ heX')⟩, eRk_insert_eq_add_one ⟨heE, heX⟩⟩
+
+lemma IsRkFinite.closure_eq_closure_of_subset_of_forall_insert (hX : M.IsRkFinite X) (hXY : X ⊆ Y)
+    (hY : ∀ e ∈ Y \ X, M.eRk (Insert.insert e X) ≤ M.eRk X) : M.closure X = M.closure Y := by
+  refine hX.closure_eq_closure_of_subset_of_eRk_ge_eRk hXY <| not_lt.1 fun hlt ↦ ?_
+  obtain ⟨z, hz, hr⟩ := exists_eRk_insert_eq_add_one_of_lt hlt
+  simpa [hr, ENat.add_one_le_iff hX.eRk_lt_top.ne] using hY z hz
+
+lemma eRk_eq_of_eRk_insert_le_forall (hXY : X ⊆ Y)
+    (hY : ∀ e ∈ Y \ X, M.eRk (insert e X) ≤ M.eRk X) : M.eRk X = M.eRk Y := by
+  refine (M.eRk_mono hXY).eq_of_not_lt <| fun hlt ↦ ?_
+  obtain ⟨e, he, hins⟩ := exists_eRk_insert_eq_add_one_of_lt hlt
+  exact (hins.symm.trans_le <| hY e he).not_lt <|
+    (ENat.lt_add_one_iff (hlt.trans_le le_top).ne).2 rfl.le
+
+/-! ### Independence -/
+
+lemma indep_iff_eRk_eq_encard_of_finite (hI : I.Finite) : M.Indep I ↔ M.eRk I = I.encard := by
+  refine ⟨fun h ↦ by rw [h.eRk_eq_encard], fun h ↦ ?_⟩
+  obtain ⟨J, hJ⟩ := M.exists_isBasis' I
+  rw [← hI.eq_of_subset_of_encard_le' hJ.subset]
+  · exact hJ.indep
+  rw [← h, ← hJ.eRk_eq_encard]
+
+lemma indep_iff_eRk_eq_encard [M.RankFinite] : M.Indep I ↔ M.eRk I = I.encard := by
+  refine ⟨Indep.eRk_eq_encard, fun h ↦ ?_⟩
+  obtain hfin | hinf := I.finite_or_infinite
+  · rwa [indep_iff_eRk_eq_encard_of_finite hfin]
+  rw [hinf.encard_eq] at h
+  exact False.elim <| (M.isRkFinite_set I).eRk_lt_top.ne h
```

Comment:
These two have identical conclusions, is there some way we can derive one from the other; or some common assumption that proves both? If not, perhaps cross-referencing could be useful to describe the difference

Reply:
I don't see a common abstraction - basically the second follows from the first because infinite sets can't satisfy the RHS or LHS in an finite-rank matroid. I've added a note.